llvm/test/Transforms/OpenMP/barrier_removal.ll

```diff
@@ -1270,77 +1270,21 @@ exit:
 ; MODULE: attributes #[[ATTR1:[0-9]+]] = { convergent nocallback nounwind }
 ; MODULE: attributes #[[ATTR2:[0-9]+]] = { convergent nocallback nofree nounwind willreturn }
 ; MODULE: attributes #[[ATTR3:[0-9]+]] = { nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) }
-; MODULE: attributes #[[ATTR4]] = { "kernel" }
-; MODULE: attributes #[[ATTR5]] = { nosync memory(none) }
+; MODULE: attributes #[[ATTR4]] = { "kernel" "nvvm.kernel" }
+; MODULE: attributes #[[ATTR5]] = { "kernel" }
+; MODULE: attributes #[[ATTR6]] = { nosync memory(none) }
 ;.
 ; CGSCC: attributes #[[ATTR0]] = { "llvm.assume"="ompx_aligned_barrier" }
 ; CGSCC: attributes #[[ATTR1:[0-9]+]] = { convergent nocallback nounwind }
 ; CGSCC: attributes #[[ATTR2:[0-9]+]] = { convergent nocallback nofree nounwind willreturn }
 ; CGSCC: attributes #[[ATTR3:[0-9]+]] = { nocallback nofree nosync nounwind willreturn memory(inaccessiblemem: write) }
-; CGSCC: attributes #[[ATTR4]] = { "kernel" }
-; CGSCC: attributes #[[ATTR5]] = { nosync memory(none) }
+; CGSCC: attributes #[[ATTR4]] = { "kernel" "nvvm.kernel" }
```

Comment:
Do we need a distinction between "kernel" and "nvvm.kernel" ?

Reply:
Unfortunately, I think we do. "kernel" is really more like "OpenMP kernel" and the semantics for this do not seem to be a perfect match for "nvvm.kernel". For example, `@multiple_blocks_functions_non_kernel_effects_2` in this test has "kernel" but is not an nvvm kernel. I'm vary unfamiliar with the OpenMP semantics so I thought keeping it separate would be the safest approach, it also may be clearest to have a common "nvvm.*" prefix for all attributes currently represented as nvvm.annotations.

Reply:
I thought OpenMP just used `nvvm.kernel` as well? As far as I was aware it just controlled whether or not the function got `.entry` in the PTX.

Reply:
The problem is that OpenMP seems to need to be able to draw a distinction between OpenMP kernels and nvvm kernels. For example here it seems like OpenMP only wants to look at "kernel" not "nvvm.kernel". As a result it seems like these attributes cannot be easily unified. 
https://github.com/llvm/llvm-project/blob/c835b48a4d72227b174bcd86f071238a1583803a/llvm/lib/Transforms/IPO/OpenMPOpt.cpp#L5932-L5938

Reply:
Yes, as the comment pointed out. Only OpenMP target regions are marked as `kernel`. Nothing else.

Reply:
Don't we also emit `nvvm.kernel` on AMDGPU kernels https://godbolt.org/z/761T64Kfj? Realistically this should be a calling convention and not an attribute, but there's a lot of historical cruft around it.

Reply:
Yes, we emit both. `kernel` must be a `nvvm.kernel`/`amdgpu_kernel` but not the other way around.

> Realistically this should be a calling convention and not an attribute

Not sure if that is better. My .02 is, it is better to be called `omp_kernel`. Adding a CC is much overkill for this identification purpose, and a CC has to be target specific, while apparently OpenMP target region is not.

Reply:
@jhuber6 can OMP and CUDA kernels ever be mixed in the same TU?

Reply:
Theoretically yes, but I don't see how that works in the current compilation pipeline. However, both CUDA and OpenMP kernels could reside in the same LLVM module at link time.

Reply:
I agree that `"omp_kernel"` seems like a much better name for the meaning we're currently signifying with the `"kernel"` attribute. 

> Realistically this should be a calling convention and not an attribute, but there's a lot of historical cruft around it.

@jhuber6 are you saying that the `"kernel"` attribute should be a calling convention? or that `"nvvm.kernel"` should be (similar to `amdgpu_kernel`)?

Reply:
From my understanding, the reason we have `amdgpu_kernel` as a CC is, we also have other kinds of "kernels" that have different calling conventions, such as `amdgpu_cs`. @arsenm knows more about this.

If `nvvm.kernel` is the only kind you have, I'm not sure how much you would gain by making it a CC.

The current `kernel` should not be a CC because it is not. It has no calling convention related meaning. It is purely an identification used to identify kernel entry points for OpenMP target regions.

Reply:
Even if you only have one CC it makes much more sense to model kernels as calling conventions. The metadata use useful if we have cases where we really want fast lookup of all the kernels in the TU.

Reply:
There is a `ptx_kernel` calling convention which is an alternative to `nvvm.annoations` `!"kernel"` already. However, I don't think we can safely auto-upgrade to this in all cases, in the openMP example @jhuber6 provided above the function has both `amdgpu_kernel` and `"nvvm.kernel"` which would not be possible with `ptx_kernel` CC. Is there any way around this? if not an attribute seems like the only option. 

> The metadata use useful if we have cases where we really want fast lookup of all the kernels in the TU.

I don't think there are any cases where we do this, there isn't even a function to traverse the metadata and find all the kernels (that I know of).  It's far more important to be able to quickly check if a function is a kernel, which the metadata solution is fairly slow for (there is a cache hacked on to try to mitigate this but that has other issues). In addition metadata should not be used to carry semantic information like this.

Reply:
Honestly, us putting `nvvm.kernel` on everything should be fixed. AFAIK it's used in some OpenMPOpt code to quickly determine kernel entry points, but could easily be replaced by a more generic name. I think that's what `"kernel"` also does? I'd need to look into it more.

Reply:
> Honestly, us putting `nvvm.kernel` on everything should be fixed.

+1

It is likely because the OpenMPGPUCodeGen for the AMDGPU part shares the same code as NVPTX and even doesn't bother to emit the attribute based on actual target.

¯\_(ツ)_/¯

Reply:
Does PTX support calling a kernel entry point as an ordinary function from another function? I've never understood how PTX doesn't set an IR calling convention for the entries

Reply:
It's an error at the PTX level so the backend just lets it happen and waits until `ptxas` blows up AFAICT.

Reply:
Yeah that really should just be a separate calling convention

Reply:
Okay, fair enough. I'll start switch us over to a calling convention in https://github.com/llvm/llvm-project/pull/120806